tests/test_grammar.py

```diff
@@ -254,16 +254,17 @@
     # Interpolations in quoted strings.
     ("str_quoted_inter", "'${null}'", "None"),
     ("str_quoted_esc_single_1", r"'ab\'cd\'\'${str}'", "ab'cd''hi"),
-    ("str_quoted_esc_single_2", r"""'\\\\\${foo}'""", r"\${foo}"),
+    ("str_quoted_esc_single_2", r"""'\\\${foo}'""", r"\${foo}"),
     ("str_quoted_esc_double_1", r'"ab\"cd\"\"${str}"', 'ab"cd""hi'),
-    ("str_quoted_esc_double_2", r'''"\\\\\${foo}"''', r"\${foo}"),
+    ("str_quoted_esc_double_2", r'''"\\\${foo}"''', r"\${foo}"),
```

Comment:
Actually the double un-escaping is gone (yay!), which is why we went down from 5 to 3 backslashes. The reason for the 3 backslashes is that:
* `\${foo}` resolves to `${foo}` (escaped interpolation)
* `\\${foo}` resolves to `\<value_of_foo>` (backslash followed by a regular interpolation)
* `\\\${foo}` resolves to `\${foo}` (backslash followed by an escaped interpolation)

Reply:
confirmed in quotes strings:
```python
In [45]: OmegaConf.create({"a": r"${ident:'abc \${b}'}", "b": 10}).a
Out[45]: 'abc ${b}'

In [46]: OmegaConf.create({"a": r"${ident:'abc \\${b}'}", "b": 10}).a
Out[46]: 'abc \\10'

In [47]: OmegaConf.create({"a": r"${ident:'abc \\\${b}'}", "b": 10}).a
Out[47]: 'abc \\${b}'
```

And even in nested quoted strings, which is awesome:
```python
In [61]: OmegaConf.create({"a": r"${ident:${ident:'abc \${b}'}}", "b": 10}).a
Out[61]: 'abc ${b}'

In [62]: OmegaConf.create({"a": r"${ident:${ident:'abc \\${b}'}}", "b": 10}).a
Out[62]: 'abc \\10'

In [63]: OmegaConf.create({"a": r"${ident:${ident:'abc \\\${b}'}}", "b": 10}).a
Out[63]: 'abc \\${b}'
```

But not in unquoted strings:
```python
In [50]: OmegaConf.create({"a": r"${ident:abc ${b}}", "b": 10}).a
Out[50]: 'abc 10'

In [51]: OmegaConf.create({"a": r"${ident:abc \${b}}", "b": 10}).a
Out[51]: 'abc \\10'

In [52]: OmegaConf.create({"a": r"${ident:abc \\${b}}", "b": 10}).a
Out[52]: 'abc \\10'

In [53]: OmegaConf.create({"a": r"${ident:abc \\\${b}}", "b": 10}).a
Out[53]: 'abc \\\\10'
```
I want to make sure this is intentional.

Reply:
It's kind of intentional, in the sense that it's a side effect of intentional features, but I agree it can be confusing. I admit that I never worried about escaped interpolations outside of quoted strings, so it's good to have this discussion.

I don't think we can actually support them in the intuitive way you may like to see, where I guess one may want the following to work (as you mentioned, it doesn't currently behave this way):
```python
In [51]: OmegaConf.create({"a": r"${ident:abc \${b}}", "b": 10}).a
Out[51]: 'abc ${b}'
```
The problem I see is that when we talk about an escaped interpolation, it's only the opening of the interpolation `\${` that is escaped: the closing brace `}` isn't. It doesn't matter in quoted strings, because closing braces aren't special characters, so we can just copy them and we don't even care for matching braces, for instance `"foo \${bar"` is a valid quoted string.
However, in unquoted strings a closing brace matters, and the example above should actually crash if `\${` was considered an escaped interpolation, because the first `}` would match the opening `${ident:` and the second one wouldn't match anything (=> GrammarParseError).

That being said, in unquoted strings `$` is a valid character and both `{` and `}` can be escaped, so what we can write is:
```python
In [51]: OmegaConf.create({"a": r"${ident:abc $\{b\}}", "b": 10}).a
Out[51]: 'abc ${b}'
```

I agree it may be a bit confusing, though it's such a niche feature that I personally don't worry too much about it. Right now I don't see a really good solution to this problem, but let me know if you do.

Reply:
> Do we have tests for all these cases (also potentially for the symmetric cases with double quotes?)

Doesn't look like it ==> added tests in 2c1f79c

Reply:
> I don't think we can actually support them in the intuitive way you may like to see, where I guess one may want the following to work (as you mentioned, it doesn't currently behave this way):
> 
> ```python
> In [51]: OmegaConf.create({"a": r"${ident:abc \${b}}", "b": 10}).a
> Out[51]: 'abc ${b}'
> ```
> 
> The problem I see is that when we talk about an escaped interpolation, it's only the opening of the interpolation `\${` that is escaped: the closing brace `}` isn't. It doesn't matter in quoted strings, because closing braces aren't special characters, so we can just copy them and we don't even care for matching braces, for instance `"foo \${bar"` is a valid quoted string.
> However, in unquoted strings a closing brace matters, and the example above should actually crash if `\${` was considered an escaped interpolation, because the first `}` would match the opening `${ident:` and the second one wouldn't match anything (=> GrammarParseError).
> That being said, in unquoted strings `$` is a valid character and both `{` and `}` can be escaped, so what we can write is:
> 
> ```python
> In [51]: OmegaConf.create({"a": r"${ident:abc $\{b\}}", "b": 10}).a
> Out[51]: 'abc ${b}'
> ```
> 
> I agree it may be a bit confusing, though it's such a niche feature that I personally don't worry too much about it. Right now I don't see a really good solution to this problem, but let me know if you do.

Here is a thought:
What if we made escapedInterpolation a grammar rule?

(There is room to refactor to reduce duplication with the unescaped interpolation case)):
```antlr
escapedInterpolation: escapedInterpolationNode | escapedInterpolationResolver;
escapedInterpolationResolver: ESC_INTER resolverName COLON sequence? BRACE_CLOSE;
escapedInterpolationNode:
      ESC_INTER
      DOT*                                                     // relative interpolation?
      (configKey | BRACKET_OPEN configKey BRACKET_CLOSE)       // foo, [foo]
      (DOT configKey | BRACKET_OPEN configKey BRACKET_CLOSE)*  // .foo, [foo], .foo[bar], [foo].bar[baz]
      INTER_CLOSE;
```

(If we decide to try it out, It should probably be in a followup diff).

Reply:
Besides the treatment inside resolvers, there is also the treatment for free strings.
I include both styles in the example below, not suggesting they they should both work (just to see what they both look like):
e.g.:
```yaml
user: Dave
# simple interpolation
a1: ${user}  # Dave
b1: hello ${user} # hello Dave

# string interpolation, style 1
a2: \${user} # ${user}
b2: hello \${user} # hello ${user}

# string interpolation, style 2
a3: $\{user\} # ${user}
b3: hello $\{user\} # hello ${user}
```

Reply:
> What if we made escapedInterpolation a grammar rule?
> 
> ```antlr
> escapedInterpolation: escapedInterpolationNode | escapedInterpolationResolver;
> escapedInterpolationResolver: ESC_INTER resolverName COLON sequence? BRACE_CLOSE;
> escapedInterpolationNode:
>       ESC_INTER
>       DOT*                                                     // relative interpolation?
>       (configKey | BRACKET_OPEN configKey BRACKET_CLOSE)       // foo, [foo]
>       (DOT configKey | BRACKET_OPEN configKey BRACKET_CLOSE)*  // .foo, [foo], .foo[bar], [foo].bar[baz]
>       INTER_CLOSE;
> ```

I guess we could, but the original goal (at least, my goal) with escaping `\${` was to make sure that someone could (easily) have a config string containing any sequence of characters. If we enforce that escaped interpolations *must* look like regular interpolations, then we lose some coverage (for instance how do you get a config string evaluating to `${`?), at which point I would suggest to just remove escaped interpolations and instead use the resolver-based approach I suggested at the end of https://github.com/omry/omegaconf/pull/630#discussion_r619170109

Another option I considered is to use what you called "style 2" in top-level & quoted strings as well (instead of just unquoted strings) -- and thus removing `\${` which wouldn't be needed anymore. I think it could work -- the only thing that bugs me is the handling of braces in top-level/quoted strings: forcing users to escape them could be a bit confusing especially in strings with no `$` sign -- why would I need to write `foo\{${bar}\}` but `foo{bar}` works? And if we don't force the escaping, then it's also a bit confusing because there are multiple ways to achieve the same result. That being said, again, I think it's doable.

Reply:
> I guess we could, but the original goal (at least, my goal) with escaping \${ was to make sure that someone could (easily) have a config string containing any sequence of characters. If we enforce that escaped interpolations must look like regular interpolations, then we lose some coverage (for instance how do you get a config string evaluating to ${?), at which point I would suggest to just remove escaped interpolations and instead use the resolver-based approach I suggested at the end of #630 (comment)

This was an initial take to start a discussion.
`${`  can potentially be represented as `$\{`.
The core of the idea is that the parser can be used to match the closing brace for an escaped interpolation.
What if we greedly consume everything until the first closing brace  (which is not escaped)?

`\${stuff_here_cannot_contain_an_unescaped_closing_brace}`
e.g: `\${foo{bar\}}`

so something like:
```anrlr
escapedInterpolation: ESC_INTER (CHAR_EXCEPT_INTER_CLOSE)* INTER_CLOSE
```

> Another option I considered is to use what you called "style 2" in top-level & quoted strings as well (instead of just unquoted strings) -- and thus removing \${ which wouldn't be needed anymore. I think it could work -- the only thing that bugs me is the handling of braces in top-level/quoted strings: forcing users to escape them could be a bit confusing especially in strings with no $ sign -- why would I need to write foo\{${bar}\} but foo{bar} works? And if we don't force the escaping, then it's also a bit confusing because there are multiple ways to achieve the same result. That being said, again, I think it's doable.

I toyed with the idea of supporting only style 2 as well.
I dislike it because the first style is so much easier to understand.

Reply:
> Now that I think of it, this looks like it could be a lexer rule as well, no?

A lexer rule will not supporting nesting escaped interpolations.

```yaml
a: \${func:\${bar}} # func("${bar}")
```

but this seems so far down the rabbit hole that I don't really think this would ever be a problem.

Reply:
> This was an initial take to start a discussion.
> `${` can potentially be represented as `$\{`.
> The core of the idea is that the parser can be used to match the closing brace for an escaped interpolation.
> What if we greedly consume everything until the first closing brace (which is not escaped)?
> 
> `\${stuff_here_cannot_contain_an_unescaped_closing_brace}`
> e.g: `\${foo{bar\}}`

Besides the potentially counter-intuitive behavior that the escaped opening doesn't match an escaped closing, the bigger issue is that I don't want to force an escaped interpolation `\${` to have a matching closing brace.
If I understand correctly, you're saying that if there is a matching closing brace then one should use `\${...}`, and if not then they should use `$\{...` This is starting to get even more confusing IMO.

I don't think this can be a lexer rule, you need some parser logic to deal with something like `\${${foo}}`.

(sorry, I still can't see a really good solution right now, I'll try to think about it more though)

Reply:
> I don't think this can be a lexer rule, you need some parser logic to deal with something like `\${${foo}}`.

This is would be parsed as the token `${${foo}` followed by the token `}`.
to make it parse as `${${foo}}`, the inner closing quote needs to be escaped:
```
\${${foo\}}
```

Reply:
> Besides the potentially counter-intuitive behavior that the escaped opening doesn't match an escaped closing, the bigger issue is that I don't want to force an escaped interpolation \${ to have a matching closing brace.

Not sure what will happen if we use a lexer rule like the above if there is no matching closing brace. we could potentially make it fallback to be a a simple string.
```
\${aaaa}  # escaped interpolation
\${aaaa  # regular string
```

> If I understand correctly, you're saying that if there is a matching closing brace then one should use \${...}, and if not then they should use $\{... This is starting to get even more confusing IMO.

I don't think I am suggesting it.

Reply:
> to make it parse as `${${foo}}`, the inner closing quote needs to be escaped:
> 
> ```
> \${${foo\}}
> ```

If we do that then we're ending up with something similar to what we used to do with quoted strings: we need to un-escape the braces in the middle of `\${...}` then run another parsing on it. I expect we will run into a similar double un-escaping problem as we are trying to solve here. Also, since what's in the middle may contain more complex interpolations (included nested ones), having to escape all the `\}` would probably make things quite hard to read.

Reply:
> To be honest, I don't think we should be sweating over the usecase of someone really wanting to have `\${` in an quoted string.

That's how I'm starting to feel as well, except I would have said "in an unquoted string", is that what you meant?

One potential idea to at least remove part of the confusion would be to raise an explicit error if someone tries to use `\${` in an unquoted string, to tell them to either use `\\${` (if they want a backslash followed by an interpolation) or `$\{` (if they want `${`). The only concern though is that it would be a breaking change.

(I gave it more thought and I can't see an ideal solution where the nice `\${` syntax would be intuitively supported in both quoted and unquoted strings)

Reply:
Let's say we make this a lexer token.
```anrlt
ESC_CLOSING_BRACE: '\}'
ESC_INTER_CHAR: ... // all legal char tokens, including ESC_CLOSING_BRACE. not including }.
ESC_INTERPOLATION : '\${' (ESC_INTER_CHAR)* '}'
```
We do need to unescape inner `\}` (and the opening `\${`) , but there is no nesting here. we only do it once as far as I can tell.
We basically need to unescape the returned token once. When would we need to re-parse the same same string more than once?

> since what's in the middle may contain more complex interpolations (included nested ones), having to escape all the \} would probably make things quite hard to read.

It will, but that is the price of having everything else so easy to read, right?
This should be a very rare scenario never needed by the majority of users.

Reply:
> Let's say we make this a lexer token.
> (...)
> We do need to unescape inner `\}` (and the opening `\${`) , but there is no nesting here. we only do it once as far as I can tell.
> We basically need to unescape the returned token once. When would we need to re-parse the same same string more than once?

If it's a lexer token then all the parser sees is this token (`ESC_INTERPOLATION`) => it can't parse its content. So the visitor has to do something like:
```python
content = ctx.getText()  # `\${${foo\}}`
content = content[3:-1]  # `${foo\}`
content = content.replace(r"\}", "}")  # `${foo}`
return escaped_interpolation_callback(content)
```
This is approximate pseudo-code, similar to what we used to have with quoted strings. I'm not entirely sure if we would also need to un-escape `\\`: this may not be the case, which would at least alleviate my double-unescaping concern. However, it does make the code more complex for a very niche feature.

There's also the issue of how to deal with non-matching braces, but I'd rather wait a bit before going into that one.

> > since what's in the middle may contain more complex interpolations (included nested ones), having to escape all the } would probably make things quite hard to read.
> 
> It will, but that is the price of having everything else so easy to read, right?

Personally I find that the quoted `"\${${foo}}"` remains easier to read than `\${${foo\}}`, so I'm not entirely sure we'd really gain something.

Reply:
> Personally I find that the quoted "\${${foo}}" remains easier to read than \${${foo\}}, so I'm not entirely sure we'd really gain something.

I agree that `\${${foo}}` is easier to read than `\${${foo\}}`. I thought that he problem with it is that requires the content inside the escaped interpolation to be parsable.
This was your argument. Personally I don't think it's a huge problem.

How are you thinking about handling cases like:
```
\${unclosed_escaped_interpolation`
\${${unclosed_escaped_custom_resolver:`
```
?

Do you prefer that we require escaped interpolations to be of style 2 instead of style 1?

Reply:
> I agree that `\${${foo}}` is easier to read than `\${${foo\}}`. I thought that he problem with it is that requires the content inside the escaped interpolation to be parsable.
> This was your argument. Personally I don't think it's a huge problem.

I'm a bit confused, with all these discussions, I can't tell exactly what "argument" you are alluding to.

My (not-so-clear) point with that last comment was that since the quoted syntax is simpler (you forgot the quotes btw in your reply), then the simpliciy argument plays in favor of keeping the current behavior (instead of trying to support a new, more complex syntax in unquoted strings).

> How are you thinking about handling cases like:
> 
> ```
> \${unclosed_escaped_interpolation
> \${${unclosed_escaped_custom_resolver:
> ```
> 
> ?
> 
> Do you prefer that we require escaped interpolations to be of style 2 instead of style 1?

My current preference would be to keep the current behavior (possibly augmented with the error message described in https://github.com/omry/omegaconf/pull/695#discussion_r622619821 though I'm not sure it's worth a breaking change). Regarding your two examples it would mean that:
* In unquoted strings, both would raise GrammarParseError
* In quoted strings, the first one would resolve to`${unclosed_escaped_interpolation` while the second one would raise a GrammarParseError

If we really want to enable the same syntax between quoted and unquoted strings, my preference would be to rely on style 2 because it is more straigthforward to implement & understand (even if a bit more cumbersome).

Reply:
> > I agree that `\${${foo}}` is easier to read than `\${${foo\}}`. I thought that he problem with it is that requires the content inside the escaped interpolation to be parsable.
> > This was your argument. Personally I don't think it's a huge problem.
> 
> I'm a bit confused, with all these discussions, I can't tell exactly what "argument" you are alluding to.
> 
> My (not-so-clear) point with that last comment was that since the quoted syntax is simpler (you forgot the quotes btw in your reply), then the simpliciy argument plays in favor of keeping the current behavior (instead of trying to support a new, more complex syntax in unquoted strings).

I misunderstood you.
> Personally I find that the quoted "\${${foo}}" remains easier to read than \${${foo\}}, so I'm not entirely sure we'd really gain something.

I thought we were talking about strings that are not quoted.
And I realize now that this can mean two things:
1. unquotes strings inside interpolations.
2. top level strings.

My concern is about top level strings. Inside interpolation it's perfectly fine to require quoting in some scenarios (this is what quoted strings exists for anyway, to short circuit some parser logic, e.g parsing of integers and dictionaries.

Here is something I would like to support (modulo any mistakes I have here, this does not work right now. this is mostly an attempt to define what I want but I could have some mistakes):
```python
from omegaconf import OmegaConf

cfg = OmegaConf.create(
    {
        "x": "V",
        "a0": r"${x}",  # V
        "a1": r"\${x}",  # ${x}
        "a2": r"Hello ${x}",  # Hello V
        "a3": r"Hello \${x}",  # Hello ${x}
        "a4": r"Hello \${{}",  # Hello ${{}
        "a5": r"Hello \${{}}",  # Hello ${{}}
        "a6": r"Hello \${{}} ${x}",  # Hello ${{}} V
        "a7": r"Hello \${broken nested ${interpolations}",  # Hello ${broken nested ${interpolations}
        "a8": r"Hello \${unclosed",  # Hello ${unclosed
        "a9": r"Hello \${foo ${x}}",  # Hello ${foo ${x}}
        # for completeness
        "a10": r"Hello \\${x}",  # Hello \${x}
    }
)

assert cfg.a0 == "V"
assert cfg.a1 == "${x}"
assert cfg.a2 == "Hello V"
assert cfg.a3 == "Hello ${x}"
# assert cfg.a4 == "Hello ${{}"
# assert cfg.a5 == "Hello ${{}}"
# assert cfg.a6 == "Hello ${{}} V"
# assert cfg.a7 == "Hello ${broken nested ${interpolations}"
# assert cfg.a8 == "Hello ${unclosed
# assert cfg.a9 == "Hello ${foo ${x}}"
# assert cfg.a10 == r"Hello \${x}"
```

Things here are going deeper and deeper into the rabbit hole.
But I think the most important things are supported already.
I tried to use the identity trick to address the unsupported cases but ran into issues that are probably due to #666:
```python

cfg = OmegaConf.create(
    {
        "x": "V",
        "a1": r"${oc.literal: '{'}",  # {
        "a2": r"${oc.literal: '{}'}",  # {}
        "a3": r"${oc.literal: '\${foo}'}",  # ${foo}
        "a4": r"${oc.literal: '\${}'}",  # ${}
        # I think all those are failing because of the issue described in #666
        # "a5": r"Hello \${{}}",  # Hello ${{}}
        # "a6": r"Hello \${{}} ${x}",  # Hello ${{}} V
        # "a7": r"Hello \${broken nested ${interpolations}",  # Hello ${broken nested ${interpolations}
        # "a8": r"Hello \${unclosed",  # Hello ${unclosed
        # "a9": r"Hello \${foo ${x}}",  # Hello ${foo ${x}}
        # # for completeness
        # "a10": r"Hello \\${x}",  # Hello \${x}
    }
)

assert cfg.a1 == "{"
assert cfg.a2 == "{}"
assert cfg.a3 == "${foo}"
# assert cfg.a4 == "${}"
```

Reply:
> I thought we were talking about strings that are not quoted.
> And I realize now that this can mean two things:
> 
> 1. unquotes strings inside interpolations.
> 2. top level strings.
> 
> My concern is about top level strings. Inside interpolation it's perfectly fine to require quoting in some scenarios (this is what quoted strings exists for anyway, to short circuit some parser logic, e.g parsing of integers and dictionaries.

So, actually, top-level strings behave the same way as quoted strings, so whenever I was talking about "unquoted strings" before, I was only thinking of the unquoted strings found in resolver arguments (e.g.: `${foo: a, [b, c, d], {e: 0, f: 1}}` where a,b,c,d,e,f are all unquoted strings).

> Here is something I would like to support (modulo any mistakes I have here, this does not work right now. this is mostly an attempt to define what I want but I could have some mistakes):
> 
> ```python
> from omegaconf import OmegaConf
> 
> cfg = OmegaConf.create(
>     {
>         "x": "V",
>         "a0": r"${x}",  # V
>         "a1": r"\${x}",  # ${x}
>         "a2": r"Hello ${x}",  # Hello V
>         "a3": r"Hello \${x}",  # Hello ${x}
>         "a4": r"Hello \${{}",  # Hello ${{}
>         "a5": r"Hello \${{}}",  # Hello ${{}}
>         "a6": r"Hello \${{}} ${x}",  # Hello ${{}} V
>         "a7": r"Hello \${broken nested ${interpolations}",  # Hello ${broken nested ${interpolations}
>         "a8": r"Hello \${unclosed",  # Hello ${unclosed
>         "a9": r"Hello \${foo ${x}}",  # Hello ${foo ${x}}
>         # for completeness
>         "a10": r"Hello \\${x}",  # Hello \${x}
>     }
> )
> 
> assert cfg.a0 == "V"
> assert cfg.a1 == "${x}"
> assert cfg.a2 == "Hello V"
> assert cfg.a3 == "Hello ${x}"
> # assert cfg.a4 == "Hello ${{}"
> # assert cfg.a5 == "Hello ${{}}"
> # assert cfg.a6 == "Hello ${{}} V"
> # assert cfg.a7 == "Hello ${broken nested ${interpolations}"
> # assert cfg.a8 == "Hello ${unclosed
> # assert cfg.a9 == "Hello ${foo ${x}}"
> # assert cfg.a10 == r"Hello \${x}"
> ```

I haven't looked closely at all of these, but most fail because of #666. I disagree on the expected output for a few of them though:
* a7: `r"Hello \${broken nested ${interpolations}`" ==> this should attempt to resolve `${interpolations}` and fail because it doesn't exist in the config
* a9: `r"Hello \${foo ${x}}"` ==> `Hello ${foo V}`
* a10: `r"Hello \\${x}"` ==> `Hello \V`

For both a7 and a9, the motivation is that `\${` just indicates that we want the sequence of characters `${`, but says nothing about what comes afterwards, which should be processed as usual.
If instead we try to identify an escaped `\${..}` block and not resolve anything in the middle, then how do we obtain a string with resolved interpolations within after `${`? For instance, here, how would I obtain the string `${V}`? (by interpolating through `x`)? (right now it works with `\${${x}}`)

For a10, how would I obtain `Hello \V` if it returned `Hello \${x}`? (right now to obtain `Hello \${x}` you need an extra backslash: `Hello \\\${x}`, i.e., an escaped backslash followed by an escaped interpolation)

> I tried to use the identity trick to address the unsupported cases but ran into issues that are probably due to #666:

I haven't checked these examples since quoted strings will indeed suffer from the same failures due to #666.

Reply:
> > I thought we were talking about strings that are not quoted.
> > And I realize now that this can mean two things:
> > 
> > 1. unquotes strings inside interpolations.
> > 2. top level strings.
> > 
> > My concern is about top level strings. Inside interpolation it's perfectly fine to require quoting in some scenarios (this is what quoted strings exists for anyway, to short circuit some parser logic, e.g parsing of integers and dictionaries.
> 
> So, actually, top-level strings behave the same way as quoted strings, so whenever I was talking about "unquoted strings" before, I was only thinking of the unquoted strings found in resolver arguments (e.g.: `${foo: a, [b, c, d], {e: 0, f: 1}}` where a,b,c,d,e,f are all unquoted strings).

Okay, this was a very large gap in my understanding and have made this discussion more confusing.
This is one thing that is important to explain when documenting the grammar.
Top level strings,  which are unquoted - actually behaving nested quoted strings is something not very intuitive.
It makes sense, but this is not something many would realize without help.
 
> > Here is something I would like to support (modulo any mistakes I have here, this does not work right now. this is mostly an attempt to define what I want but I could have some mistakes):

```
...
```
> I haven't looked closely at all of these, but most fail because of #666. I disagree on the expected output for a few of them though:
> 
> * a7: `r"Hello \${broken nested ${interpolations}`" ==> this should attempt to resolve `${interpolations}` and fail because it doesn't exist in the config
> * a9: `r"Hello \${foo ${x}}"` ==> `Hello ${foo V}`
> * a10: `r"Hello \\${x}"` ==> `Hello \V`
> 
> For both a7 and a9, the motivation is that `\${` just indicates that we want the sequence of characters `${`, but says nothing about what comes afterwards, which should be processed as usual.
> If instead we try to identify an escaped `\${..}` block and not resolve anything in the middle, then how do we obtain a string with resolved interpolations within after `${`? For instance, here, how would I obtain the string `${V}`? (by interpolating through `x`)? (right now it works with `\${${x}}`)

I am okay with that interpretation.

> For a10, how would I obtain `Hello \V` if it returned `Hello \${x}`? (right now to obtain `Hello \${x}` you need an extra backslash: `Hello \\\${x}`, i.e., an escaped backslash followed by an escaped interpolation)

I am okay with interpretation as well. There is no need to diverge than what quoted string is doing.

Reply:
> By the way, I think we are still not exactly aligned with the grammar definitions here.
> In the grammar, topLevelString appears to be the stuff between interpolations.
> I am thinking of the whole value as a top level string:
> 
> ```
> thing: string
> x: I am a top level ${thing}
> #  ^^^^^^^^^^^^^^^^^^^^^^^^^
> ```
> 
> I think you are only considering `I am a top level ` as a top level string.

That's correct. If you find a better name let me know!

> I am okay with that interpretation.
> (...)
> I am okay with interpretation as well. There is no need to diverge than what quoted string is doing.

So (ignoring #666 for now) what are your updated thoughts on trying to support escaped interpolations in *unquoted* strings? (i.e., to be able to write e.g. `${foo:abc\${bar}}`)

Reply:
> That's correct. If you find a better name let me know!


I think topLevelStr can also be called simpleText, IE : text without interpolations.

The bad name made me thing that those are specific to first level of the parsing (directly in the value).
Looking at the parser, this is indeed the case now but I think the grammar can be refactored a bit to make reuse simpleText as both inside quoted string and inside configValue.

something like this:

```antlr
configValue: text EOF;

# former toplevelStr, text without interpolation
simpleText: (ESC | ESC_INTER | SPECIAL_CHAR | ANY_STR)+;

# text with interpolations
text: (simpleText | (simpleText? (interpolation simpleText?)+));

quotedValue: QUOTE_OPEN text QUOTE_CLOSE;
```

Note that it's much easier to understand the relation between configValue and quotedValue in this form.

I didn't try this so it it may not be doable.

You have other TOP this and TOP that lexer which I didn't think about.

> So (ignoring #666 for now) what are your updated thoughts on trying to support escaped interpolations in unquoted strings? (i.e., to be able to write e.g. ${foo:abc\${bar}})

I think it's not worth the effort to add specific support for it given that we have an easy to quote the string here:
```
${foo:'abc\${bar}'}
```
All of this should be clearly documented in some corner of the docs though (I know there is an item to document the grammar).

Reply:
> Looking at the parser, this is indeed the case now but I think the grammar can be refactored a bit to make reuse simpleText as both inside quoted string and inside configValue.

I think that's a great idea! Done in 8550d72. Then I did a few follow-ups as well:
* 1779e26 => minor refactor to keep the methods sorted
* 1b91270 => simplification of the grammar as well as the visitor code (I don't remember exactly why I had this somewhat complex definition for text/topLevelStr -- there may or may not have been a good reason before, but now I don't see any)
* 52eebfe => suggestion to simplify the `primitive` rule (I don't feel strongly about that one, happy to revert if you preferred the previous version)

> I think it's not worth the effort to add specific support for it given that we have an easy to quote the string here:
> 
> ```
> ${foo:'abc\${bar}'}
> ```

Agreed :)

> All of this should be clearly documented in some corner of the docs though (I know there is an item to document the grammar).

Agreed as well!